plugins/woocommerce/src/Internal/Admin/ProductDownloadsPreview.php

```diff
@@ -0,0 +1,240 @@
+				'woocommerce_rest_file_not_found',
+				__( 'File not found', 'woocommerce' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		// Only allow image mime types.
+		$mime_type = get_post_mime_type( $attachment_id );
+		$allowed_mime_types = array(
+			'image/jpeg',
+			'image/jpg',
+			'image/png',
+			'image/gif',
+			'image/webp',
+		);
+
+		if ( ! in_array( $mime_type, $allowed_mime_types, true ) ) {
+			return new WP_Error(
+				'woocommerce_rest_invalid_file_type',
+				__( 'Invalid file type', 'woocommerce' ),
+				array( 'status' => 403 )
+			);
+		}
+
+		// Handle image size requests - use 'large' as maximum size, 'full' can be too large.
+		$size = $requested_size === 'full' ? 'large' : $requested_size;
+
+		$resized = image_get_intermediate_size( $attachment_id, $size );
+		if ( $resized && isset( $resized['path'] ) ) {
+			$uploads_dir = wp_upload_dir();
+			$resized_file_path = $uploads_dir['basedir'] . '/' . $resized['path'];
+			if ( is_readable( $resized_file_path ) ) {
+				$file_path = $resized_file_path;
+			}
+		}
+
+		// Serve the file with appropriate headers.
+		$this->clean_buffers();
+
+		// We need to manually build the response to send proper binary data.
+		nocache_headers();
+		header( 'Content-Type: ' . $mime_type );
+		header( 'Content-Length: ' . filesize( $file_path ) );
+		header( 'Content-Disposition: inline; filename="' . basename( $file_path ) . '"' );
+
+		readfile( $file_path );
+		exit;
+	}
+
+	/**
+	 * Get secure URL for admin image that works with image src attributes
+	 *
+	 * @param int    $product_id    Product ID.
+	 * @param int    $attachment_id Attachment ID.
+	 * @param string $size          Image size.
+	 * @return string Secure admin image URL.
+	 *
+	 * @since 9.9.0
+	 */
+	public function get_admin_image_src_url( $product_id, $attachment_id, $size ) {
+		if ( ! current_user_can( 'manage_woocommerce' ) ) {
+			return '';
+		}
+
+		// Generate a secure token.
+		$token = wp_generate_password( 32, false );
+		
+		// Store token in transient with 15-minute expiration.
+		$transient_key = 'wc_preview_token_' . $token;
+		$token_data = [
+			'attachment_id' => $attachment_id,
+			'product_id' => $product_id,
+		];
+		
+		set_transient( $transient_key, $token_data, 15 * MINUTE_IN_SECONDS );
+		
+		// Generate URL with token.
+		$url = rest_url( "wc/v3/admin/product-downloads-preview/{$product_id}/{$attachment_id}" );
+		$url = add_query_arg(
+			[
+				'size' => $size,
+				'token' => $token,
+			],
+			$url
+		);
+		
+		return $url;
+	}
+
+	/**
+	 * Clean all output buffers.
+	 *
+	 * Can prevent errors, for example: transfer closed with 3 bytes remaining to read.
+	 */
+	private function clean_buffers() {
+		if ( ob_get_level() ) {
+			$levels = ob_get_level();
+			for ( $i = 0; $i < $levels; $i++ ) {
+				@ob_end_clean(); // phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+			}
+		} else {
+			@ob_end_clean(); // phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+		}
```

Comment:
Do we need the `else` clause? If there are no buffers (if `ob_get_level()` returns `(int) 0`, which I think is the only condition that will lead us into this path), there is nothing to clean.

Reply:
No need for all the complications here, our function is the only one on that endpoint and so I've opted for something much simpler in: fe79a6c3c3af5a873be8857084169d4dc3b49cc1